Upgradable contracts using `set_code` function

yarikbratashchuk opened the pull request:
This pr is a follow up for #10567.  

The question that naturally comes to mind: 
- Do we need to check if there is deployed contract behind the hash that is being set?

That would give us a little bit for of certainty when calling `set_code` function. On the other hand we won't be able to set hash for the contract that has not been deployed yet. As for me it doesn't make much sense considering that we don't make full check of what we're setting with this call anyways.

Thoughts and comments are required and welcome.

athei commented:
I think we need to check whether the code hash already exists. It is necessary because we need to increase the ref count of the code that we are switching to. This will prevent the code from being deleted. The reason for that is that we don't want the owner of a code being able to delete it when it is in use.

We could of course get rid of the ref counting and let the owner always delete the code. Anyone who still needs the code would then need to re-upload the code (they can get it from an archive node). However, this can lead to intermediate breakage and might prevent people from sharing code.

yarikbratashchuk commented:
Makes sense. Will add the check using `contains_storage` method.

yarikbratashchuk commented:
@athei, please take a look at what has been done, and let me know what you think

athei commented on frame/contracts/src/wasm/code_cache.rs:
```diff
@@ -118,6 +118,16 @@ pub fn decrement_refcount<T: Config>(code_hash: CodeHash<T>) -> Result<(), Dispa
 	Ok(())
 }
 
+/// Increment the refcount of a code in-storage by one.
+pub fn increment_refcount<T: Config>(code_hash: CodeHash<T>) -> Result<(), DispatchError> {
+	<OwnerInfoOf<T>>::mutate(code_hash, |existing| {
+		if let Some(info) = existing {
+			info.refcount = info.refcount.saturating_add(1);
+		}
+	});
+	Ok(())
+}
```
Make it return an error if the entry doesn't exist ( `existing = None`). Then we can use that as an error condition when setting the new code. Please make sure to document that behavior.

The return type and comment on `decrement_refcount` is wrong by the way. I forgot to update it. It is infallible and doesn't delete anything when it drops to zero. Just so you don't get confused.

yarikbratashchuk replied:
Making update

Won't touch `decrement_refcount` right now, as it is used elsewhere as well

athei commented on frame/contracts/src/exec.rs:
```diff
@@ -1016,6 +1026,19 @@ where
+	fn set_code_hash(&mut self, hash: CodeHash<Self::T>) -> Result<(), DispatchError> {
+		let top_frame = &mut self.top_frame_mut();
+		let prev_hash = top_frame.contract_info().code_hash.clone();
+		if !Self::code_exists(&hash) {
+			return Err(DispatchError::CannotLookup)
+		}
```
Don't do this check explicitly. Do the changes to `increment_refcount` I pointed out and then error out when it fails. Also, move those refcount operations up so we can error out earlier.

athei commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -406,6 +409,10 @@ mod tests {
 			}
 			Ok(result)
 		}
+		fn set_code_hash(&mut self, hash: CodeHash<Self::T>) -> Result<(), DispatchError> {
+			self.storage.insert(StorageKey::from(hash), vec![1, 2, 3]);
+			Ok(())
+		}
```
This feels hacky. We should add a new field to the mock ext which records all set code hashes (just like we do for other stuff).

athei commented on frame/contracts/src/benchmarking/mod.rs:
```diff
@@ -808,6 +808,46 @@ benchmarks! {
 		let origin = RawOrigin::Signed(instance.caller.clone());
 	}: call(origin, instance.addr, 0u32.into(), Weight::MAX, None, vec![])
 
+	seal_set_code_hash {
```
This is needed to prevent spamming of the output file because we create a lot of unique storage items.

```suggestion
    #[skip_meta]
	seal_set_code_hash {
```

athei commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -813,6 +818,24 @@ define_env!(Env, <E: Ext>,
 		ctx.set_storage(key_ptr, value_ptr, value_len).map(|_| ())
 	},
 
+	// Replace contract code
```
```suggestion
	// Replace contract code.
```

athei commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -813,6 +818,24 @@ define_env!(Env, <E: Ext>,
+	//
+	// # Errors
+	//
+	// - requested buffer is not within the bounds of the sandbox memory.
+	// - the buffer contents cannot be decoded as the required type.
```
Those are implicit. We don't document them here. Also, those aren't errors but panics.

athei commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -813,6 +818,24 @@ define_env!(Env, <E: Ext>,
+	// - requested buffer is not within the bounds of the sandbox memory.
+	// - the buffer contents cannot be decoded as the required type.
+	// - code for specified hash does not exist
+	[__unstable__] seal_set_code_hash(ctx, code_hash_ptr: u32) -> ReturnCode => {
```
`ReturnCode` isn't used. This function doesn't return anything but a `Success` of it returns `Ok`.

HCastano replied:
@yarikbratashchuk not sure if this has been addressed either

athei replied:
This was addressed. It calls `err_into_return_code` which returns `CodeNotFound` on err.

athei commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -813,6 +818,24 @@ define_env!(Env, <E: Ext>,
+	//
+	// - requested buffer is not within the bounds of the sandbox memory.
+	// - the buffer contents cannot be decoded as the required type.
+	// - code for specified hash does not exist
```
This is a panic and not an error with the current code.

yarikbratashchuk commented:
@athei, please take a look at updated.
I'm not sure how to stop execution of the contract though, what are your thoughts on this?

athei commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -813,6 +818,27 @@ define_env!(Env, <E: Ext>,
+	// # Errors
+	//
+	// `ReturnCode::CodeNotFound`
+	[__unstable__] seal_set_code_hash(ctx, code_hash_ptr: u32) -> ReturnCode => {
```
Please move it to the bottom of the macro invocation. It doesn't really fit between the different `set_storage` versions.

HCastano replied:
@yarikbratashchuk can you address this?

HCastano replied:
@yarikbratashchuk can you please move this, or justify why you're leaving it in-between the different `set_storage` versions?

athei commented on frame/contracts/src/lib.rs:
```diff
@@ -559,6 +559,14 @@ pub mod pallet {
 		/// A code with the specified hash was removed.
 		CodeRemoved { code_hash: T::Hash },
+
+		/// Contract code has been updated to one specified with code_hash
+		ContractCodeUpdated {
+			/// The contract that has been updated.
+			contract: T::AccountId,
+			/// New code hash that was set for the contract
+			code_hash: T::Hash,
+		},
```
```suggestion
		/// A contract has changed its code.
		ContractCodeUpdated {
			/// The contract that has been updated.
			contract: T::AccountId,
			/// New code hash that was set for the contract.
			code_hash: T::Hash,
		},
```

athei commented on frame/contracts/src/exec.rs:
```diff
@@ -155,6 +156,9 @@ pub trait Ext: sealing::Sealed {
 		take_old: bool,
 	) -> Result<WriteOutcome, DispatchError>;
 
+	/// Sets new code hash for existing contract
```
```suggestion
	/// Sets new code hash for existing contract.
```

athei commented on frame/contracts/src/exec.rs:
```diff
@@ -1016,6 +1020,19 @@ where
+		let top_frame = &mut self.top_frame_mut();
+		let prev_hash = top_frame.contract_info().code_hash.clone();
+		top_frame.contract_info().code_hash = hash;
+		increment_refcount::<Self::T>(hash)?;
```
Move this fallible operation to the top so we can error out earlier.

athei commented on frame/contracts/src/exec.rs:
```diff
@@ -1016,6 +1020,19 @@ where
 	}
 
+	fn set_code_hash(&mut self, hash: CodeHash<Self::T>) -> Result<(), DispatchError> {
+		let top_frame = &mut self.top_frame_mut();
```
It is already a ref.

```suggestion
		let top_frame = self.top_frame_mut();
```

athei commented on frame/contracts/src/wasm/code_cache.rs:
```diff
@@ -118,6 +118,22 @@ pub fn decrement_refcount<T: Config>(code_hash: CodeHash<T>) -> Result<(), Dispa
+///
+/// # Errors
+///
+/// DispatchError::CannotLookup is returned if there is no info for specified code_hash
```
```suggestion
/// `Error::CodeNotFound` is returned if the specified `code_hash` does not exist.
```

yarikbratashchuk replied:
Ahh, sorry

athei reviewed (dismissed):
Gets better. Regarding the termination: I need to think about that and need to come back to you.

athei commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -813,6 +818,27 @@ define_env!(Env, <E: Ext>,
+	//
+	// # Parameters
+	//
+	// - code_hash_ptr: a pointer to the buffer that contains new contract code
```
```suggestion
	// - code_hash_ptr: A pointer to the buffer that contains the new code hash.
```

yarikbratashchuk commented:
@athei, do you have any thoughts on this one? I think we could just mention in docs that tx revert scenario you talked about previously. What do you think?

athei commented:
Yeah I think we can leave it the way it is: Non terminating. More docs is always nice. It is nothing out of the ordinary, though: If you revert everything gets reset.

One other thing is bugging me: This will weaken our deterministic address assumption. Before this change you can always confidently assume that a contract address identifies a specific code hash. After this change this is no longer true for contracts which include `set_code`. I mean this might be okay because the assumption still holds for contracts which don't use `set_code`.

athei commented:
> What are the negative consequences of breaking this deterministic address assumption? 

I was hoping you had some idea since it was your proposal to add this function. Deterministic addresses are nice because you can know the address and embed it into a contract before the target needs to be actually deployed. However, this is still the case for contracts which don't use `set_code` (which are call contracts as of right now).

> We won't be able to revert these changes before it's sealed to stable, will we?

I don't understand what you mean.

yarikbratashchuk commented:
> Deterministic addresses are nice because you can know the address and embed it into a contract before the target needs to be actually deployed.

Yeah, we sacrifice this for upgradability. That was my way of thinking.

>> We won't be able to revert these changes before it's sealed to stable, will we?

>I don't understand what you mean.

I mean that we can roll out this change as an unstable feature first, and if we realise something that we are missing right now we can just revert this change, right? (In case something really bad happens, which is unlikely)

athei commented:
Yes this is correct. We want to have it as unstable first at least until we have support it in ink! and people get a chance to play with it.

athei commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -820,6 +820,11 @@ define_env!(Env, <E: Ext>,
 
 	// Replace contract code.
 	//
+	// NOTE: Avoid any operations after the `seal_set_code_hash` call, otherwise if contract
+	// calls into itself after changing its code, the new call would use the new code but when
+	// the original caller panics after returning from the sub call it would revert the changes
+	// made by `seal_set_code_hash` and the next caller would use the old code.
```
We shouldn't prescribe what to avoid. If we wanted anyone to avoid it we would have made the call terminating. We should rather only inform the user of this maybe unexpected side effect.

athei commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -813,6 +818,32 @@ define_env!(Env, <E: Ext>,
 	},
 
+	// Replace contract code.
+	//
+	// NOTE: If contract calls into itself after changing its code, the new call would use
```
```suggestion
	//
	// # NOTE
	//
	// If contract calls into itself after changing its code, the new call would use
```

athei commented on frame/contracts/src/tests.rs:
```diff
@@ -2932,3 +2932,63 @@ fn code_rejected_error_works() {
+				.unwrap();
+		assert_return_code!(result, 2);
+
+		// Checking for the last event only
+		assert_eq!(
+			System::events()[13],
```
It isn't really the last event, isn't it?

yarikbratashchuk replied:
It's the last one from what I've seen

yarikbratashchuk replied:
Double checked.

athei replied:
I expected some event related to storage deposits coming last. But your test contracts do not create or remove storage so no events are generated there. It's fine.

athei commented on frame/contracts/src/benchmarking/mod.rs:
```diff
@@ -821,6 +821,46 @@ benchmarks! {
+				},
+			],
+			call_body: Some(body::repeated_dyn(r * API_BENCHMARK_BATCH_SIZE, vec![
+				Counter(0 as u32, code_hash_len as u32), // code_hash_ptr
```
```suggestion
				Counter(0, code_hash_len as u32), // code_hash_ptr
```

athei commented on frame/contracts/src/benchmarking/mod.rs:
```diff
@@ -821,6 +821,46 @@ benchmarks! {
+			}],
+			data_segments: vec![
+				DataSegment {
+					offset: 0 as u32,
```
```suggestion
					offset: 0,
```

athei commented on frame/contracts/src/lib.rs:
```diff
@@ -564,6 +564,14 @@ pub mod pallet {
 		/// A code with the specified hash was removed.
 		CodeRemoved { code_hash: T::Hash },
+
+		/// A contract code has changed its code.
```
```suggestion
		/// A contract's code was updated.
```

seunlanlege commented:
I think only root calls should be allowed to call this function seeing as contracts are a public utility, users of a contract wouldn't like if the code changes under them.

yarikbratashchuk commented:
@seunlanlege, it's a good point actually. Let's leave it as it is for now, and if we receive any negative feedback before sealing it to stable we can think of adding this feature, I feel like making it right now is premature optimization, imho.

HCastano commented:
Will take a look tomorrow!

athei commented:
> I think only root calls should be allowed to call this function seeing as contracts are a public utility, users of a contract wouldn't like if the code changes under them. But this is more of a philosophical debate.

This was discussed in: https://github.com/paritytech/substrate/issues/10567

The bottom line is that the trust that is required for proxy contracts and a contract that uses `set_code` is the same. Proxy contracts can also change the code that is executed by changing the contract to `delegate_call` into. Users can decide to not trust contracts which use `set_code` in the same way the might not trust proxy contracts.

HCastano commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -303,6 +306,7 @@ mod tests {
 	}
 
 	pub struct MockExt {
+		code_hashes: Vec<CodeHash<Test>>,
```
Again, small thing but I'd add this at the end of the struct

HCastano commented on frame/contracts/src/wasm/code_cache.rs:
```diff
@@ -117,6 +117,22 @@ pub fn decrement_refcount<T: Config>(code_hash: CodeHash<T>) -> Result<(), Dispa
+///
+/// # Errors
+///
+/// `Error::CodeNotFound` is returned if the specified `code_hash` does not exist.
```
I'd like to the docs here if possible

HCastano commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -801,6 +806,34 @@ define_env!(Env, <E: Ext>,
 		ctx.set_storage(key_ptr, value_ptr, value_len).map(|_| ())
 	},
 
+	// Replace contract code.
```
I think this kinda undersells what's happening here. I'd add some more details here, and implications (e.g, that whole discussion you and Alex had about deterministic addressed)

yarikbratashchuk replied:
It has some description under the `# NOTE` section. I'll do some updates.

HCastano commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -801,6 +806,34 @@ define_env!(Env, <E: Ext>,
+	//
+	// # NOTE
+	//
+	// If contract calls into itself after changing its code, the new call would use
```
```suggestion
	// If a contract calls into itself after changing its code the new call would use
```

HCastano commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -801,6 +806,34 @@ define_env!(Env, <E: Ext>,
+	// # NOTE
+	//
+	// If contract calls into itself after changing its code, the new call would use
+	// the new code but when the original caller panics after returning from the sub call it
```
```suggestion
	// the new cod. However, if the original caller panics after returning from the sub call it
```

HCastano reviewed (commented):
Some small things to iron out

athei commented on frame/contracts/src/exec.rs:
```diff
@@ -1092,6 +1096,19 @@ where
+		increment_refcount::<Self::T>(hash)?;
+		let top_frame = self.top_frame_mut();
+		let prev_hash = top_frame.contract_info().code_hash.clone();
+		top_frame.contract_info().code_hash = hash;
+		decrement_refcount::<Self::T>(prev_hash)?;
```
It does not matter for correctness because we roll back on error anyways but we should do it never the less as erroring out as early as possible is still the right thing to do. As a matter of fact I did already suggest that but it was glossed over.

yarikbratashchuk replied:
@athei, sorry I forgot to move up decrement_refcount. Done

athei commented on frame/contracts/src/benchmarking/mod.rs:
```diff
@@ -821,6 +821,46 @@ benchmarks! {
+			.collect::<Result<Vec<_>, &'static str>>()?;
+		let code_hash_len = code_hashes.get(0).map(|x| x.encode().len()).unwrap_or(0);
+		let code_hashes_bytes = code_hashes.iter().flat_map(|x| x.encode()).collect::<Vec<_>>();
+		let code_hashes_len = code_hashes_bytes.len();
```
Unused variable.

yarikbratashchuk replied:
Easy to miss things here. Do you have any special tricks or setup for working with this macro?

athei replied:
Nope. Just reading the code. I think the benchmarking macro will become a attribute macro eventually.

athei commented on frame/contracts/src/weights.rs:
```diff
@@ -508,6 +509,17 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
+	// Storage: Contracts ContractInfoOf (r:1 w:1)
+	// Storage: Contracts CodeStorage (r:1 w:0)
+	// Storage: Timestamp Now (r:1 w:0)
+	fn seal_set_code_hash(r: u32, ) -> Weight {
+		   (265_861_000 as Weight)
+			   // Standard Error: 1_212_000
+			   .saturating_add((303_302_000 as Weight).saturating_mul(r as Weight))
+			   .saturating_add(T::DbWeight::get().reads(4 as Weight))
+			   .saturating_add(T::DbWeight::get().writes(1 as Weight))
+	}
```
The benchmark is wrong. It does not report any storage access (for `r`). Reason is that the code hashes do not exist which you are setting to.

athei commented on frame/contracts/src/benchmarking/mod.rs:
```diff
@@ -821,6 +821,46 @@ benchmarks! {
+		});
+		let instance = Contract::<T>::new(code, vec![])?;
+		let origin = RawOrigin::Signed(instance.caller.clone());
+	}: call(origin, instance.addr, 0u32.into(), Weight::MAX, None, vec![])
```
Please add a verify block that checks whether the contracts were really switched to the new code hash.

athei reviewed (changes requested):
Benchmark need fixing.

yarikbratashchuk commented on frame/contracts/src/lib.rs:
```diff
@@ -564,6 +564,14 @@ pub mod pallet {
 		CodeRemoved { code_hash: T::Hash },
+
+		/// A contract's code was updated.
+		ContractCodeUpdated {
```
Good point.

HCastano commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -856,6 +861,37 @@ define_env!(Env, <E: Ext>,
+	// Contracts using this API can't be assumed as having deterministic (you can know the
+	// address before the contract is actually deployed) addresses.
+	//
+	// # NOTE
```
```suggestion
	// # Note
```

HCastano commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -856,6 +861,37 @@ define_env!(Env, <E: Ext>,
 		ctx.set_storage(key_ptr, value_ptr, value_len).map(|_| ())
 	},
 
+	// Replace contract code.
+	//
+	// Contracts using this API can't be assumed as having deterministic (you can know the
+	// address before the contract is actually deployed) addresses.
+	//
+	// # NOTE
+	//
+	// If a contract calls into itself after changing its code the new call would use
+	// the new code. However, if the original caller panics after returning from the sub call it
+	// would revert the changes made by `seal_set_code_hash` and the next caller would use
+	// the old code.
```
```suggestion
	// Replace the contract code at the specified address with new code.
	//
	// # Note
	//  
	// There are a couple of important considerations which must be taken into account when
	// using this API: 
	// 
	// 1. The storage at the code address will remain untouched. This means that contract developers
	// must ensure that the storage layout of the new code is compatible with that of the old code.
	//
	// 2. Contracts using this API can't be assumed as having deterministic addresses. Said another way,
	// when using this API you lose the guarantee that an address always identifies a specific code hash.
	//
	// 3. If a contract calls into itself after changing its code the new call would use
	// the new code. However, if the original caller panics after returning from the sub call it
	// would revert the changes made by `seal_set_code_hash` and the next caller would use
	// the old code.
```

athei commented on frame/contracts/src/exec.rs:
```diff
@@ -1015,6 +1019,20 @@ where
 	}
 
+	fn set_code_hash(&mut self, hash: CodeHash<Self::T>) -> Result<(), DispatchError> {
+		increment_refcount::<Self::T>(hash)?;
```
Is the turbo fish here really necessary? Looks like it should be able to be inferred from the argument. Same for the other occurrences in this function.

athei reviewed (commented):
Please just move the new api function, tests and bechmarks to the **bottom**. New functions should be added to the bottom unless it is a new version or tight

yarikbratashchuk commented:
Okay, so I changed the benchmark and added verify block, didn't get any errors. Seems to be pretty straightforward. @athei, @HCastano what do you think?

athei commented:
You changed the benchmark so that you set it to the same code multiple times. This won't work out because this will repeatedly read from the same `OwnerInfo` and hence will still yield too few reads. You need to keep the original structure of setting it to different code hashes.

yarikbratashchuk commented:
@athei, this is what i'm getting on my machine (without #[skip_meta]):
```
        // Storage: System Account (r:1 w:0)
        // Storage: Contracts ContractInfoOf (r:1 w:1)
        // Storage: Contracts CodeStorage (r:1 w:0)
        // Storage: Timestamp Now (r:1 w:0)
        // Storage: Contracts OwnerInfoOf (r:36 w:36)
        fn seal_set_code_hash(r: u32, ) -> Weight {
                (0 as Weight)
                        // Standard Error: 7_683_000
                        .saturating_add((1_705_926_000 as Weight).saturating_mul(r as Weight))
                        .saturating_add(T::DbWeight::get().reads((99 as Weight).saturating_mul(r as Weight)))
                        .saturating_add(T::DbWeight::get().writes((99 as Weight).saturating_mul(r as Weight)))
        }
```
Seems pretty crazy :)

athei commented on frame/contracts/src/benchmarking/mod.rs:
```diff
@@ -877,7 +877,6 @@ benchmarks! {
 		let origin = RawOrigin::Signed(instance.caller.clone());
 	}: call(origin, instance.addr, 0u32.into(), Weight::MAX, None, vec![])
 
-	#[skip_meta]
```
Why change an existing benchmark?

yarikbratashchuk replied:
this #[skip_meta] relates to the seal_set_code_hash, probably github diffing messed it up. It relates to [6f925a9e64924f4152f8fa5fcf4fd9c2b8373794](https://github.com/paritytech/substrate/pull/10690/commits/6f925a9e64924f4152f8fa5fcf4fd9c2b8373794)

athei replied:
Wrong. You PR removes it from `seal_set_storage_per_new_kb`. Manually compared your branch and master.

athei commented on frame/contracts/src/benchmarking/mod.rs:
```diff
@@ -1971,6 +1970,47 @@ benchmarks! {
 		let origin = RawOrigin::Signed(instance.caller.clone());
 	}: call(origin, instance.addr, 0u32.into(), Weight::MAX, None, vec![])
 
+	#[skip_meta]
```
We don't need it here. We write many times to the same map which does not spam.

yarikbratashchuk replied:
Updated

athei reviewed (commented):
We can run the bot once the skip meta situation is sorted.

athei commented:
/benchmark runtime pallet pallet_contracts

joao-paulo-parity commented:
Weights for b32e4ec8f71a68184264e514446597e800e7e0fe are at https://gist.github.com/joao-paulo-parity/374c56737868015b7fcd3b3bb7e4bbaa

yarikbratashchuk commented:
Some crazy numbers here.

athei commented:
> Some crazy numbers here.

Which numbers do you mean exactly?

yarikbratashchuk commented:
> Which benchmark numbers do you mean exactly?
Number of reads and writes for contracts OwnerInfoOf.

athei commented:
> > Which benchmark numbers do you mean exactly?
> 
> Number of reads and writes for contracts OwnerInfoOf.

The metadata looks strange indeed. But this is only relevant for actual dispatchables. For us it is more of a debugging help. 

@shawntabrizi Do you have any idea how this quirky number of `36` is here. I thought read and writes depending on a component are not part of the metadata?

```rust
// Storage: System Account (r:1 w:0)
// Storage: Contracts ContractInfoOf (r:1 w:1)
// Storage: Contracts CodeStorage (r:1 w:0)
// Storage: Timestamp Now (r:1 w:0)
// Storage: Contracts OwnerInfoOf (r:36 w:36)
fn seal_set_code_hash(r: u32, ) -> Weight {
	(0 as Weight)
		// Standard Error: 2_158_000
		.saturating_add((932_937_000 as Weight).saturating_mul(r as Weight))
		.saturating_add(T::DbWeight::get().reads((99 as Weight).saturating_mul(r as Weight)))
		.saturating_add(T::DbWeight::get().writes((99 as Weight).saturating_mul(r as Weight)))
}
```

athei commented:
bot merge
